### PR TITLE
i1358 - restore using rsync

### PR DESCRIPTION
**Pull request by hillalex**
Not ready for review.

First review and merge: 
https://github.com/vimc/bb8/pull/5

### CODE REVIEW

**Comment by hillalex** [comment]
Okay, now ready for review

File: docker_rsync.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    starport = settings.starport
+    mounted_volume = join("/", local_volume)
+    volumes = create_volumes(settings, local_volume, "rw")
+
+    remote_path = "{}@{}:{}{}/".format(starport["user"],
+                                       starport["addr"],
+                                       starport['backup_location'],
+                                       local_volume)
+
+    run_rsync(volumes, remote_path, mounted_volume, False)
```
Comment (hillalex) [inline docker_rsync.py]: So here we want to run another command to give the `bb8` user ownership of the mounted directory - but how do I 'mount' the bb8 user into the docker container?

File: docker_rsync.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    if relative:
+        cmd = ["rsync", "-rv", "-e", "ssh", "--relative", from_path, to_path, "--delete"]
+
+    container = client.containers.run("instrumentisto/rsync-ssh", command=cmd, volumes=volumes,
+                                      detach=True)
+
+    log_from_docker(container)
+
+
+def create_volumes(settings, local_volume, volume_mode):
```
Comment (MartinEden) [inline docker_rsync.py]: This is maybe a somewhat misleading function name - it doesn't create volumes, it returns the args that the docker API takes to create them. Plus, presumably the volumes are not always created? After the first time they are just being mapped into the container.

So maybe `get_volumes_args`?

File: docker_rsync.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    mounted_volume = join("/", local_volume)
+    return {abspath(settings.known_hosts_path): {"bind": "/root/.ssh/known_hosts", "mode": "rw"},
```
Comment (MartinEden) [inline docker_rsync.py]: I'm finding this a bit hard to read. Could we maybe:

1. Space it out, so the outermost dictionary gets a line for each of its braces?
1. Move the `abspath` call into settings? Do we ever need them as relative paths?
1. Replace `"{}".format(mounted_volume)` with just `mounted_volume`

File: docker_rsync.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+
+# local_volume can be an absolute path or a named volume
+def backup_volume(settings, local_volume):
+    starport = settings.starport
+    volumes = create_volumes(settings, local_volume, "ro")
+
+    destination_path = "{}@{}:{}".format(starport["user"],
+                                         starport["addr"],
+                                         starport["backup_location"])
```
Comment (MartinEden) [inline docker_rsync.py]: Can you write this as:

```
destination_path = "{user}@{addr}:{backup_location}".format(**starport)
```

File: docker_rsync.py (hunk 1 above)
Comment (MartinEden) [reply docker_rsync.py]: I think you just run rsync with `docker run -u UID`, having first got the user ID of the bb8 user. In bash you do that with `id -u bb8`, but there may be a Python library function that will do it more nicely without shelling out.

File: docker_rsync.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+def restore_volume(settings, local_volume):
+    starport = settings.starport
+    mounted_volume = join("/", local_volume)
+    volumes = create_volumes(settings, local_volume, "rw")
+
+    remote_path = "{}@{}:{}{}/".format(starport["user"],
+                                       starport["addr"],
+                                       starport['backup_location'],
+                                       local_volume)
```
Comment (MartinEden) [inline docker_rsync.py]: Also here, to use the alternative `format` style.

**Review by MartinEden — CHANGES_REQUESTED** [review]
This is looking good on a preliminary look. I will have another think when I'm fresh on Monday

File: docker_rsync.py (hunk 1 above)
Comment (richfitz) [reply docker_rsync.py]: This should do the trick:

```python
os.getuid()
```

File: targets.py [hunk 6]
```diff
@@ -27,3 +28,8 @@ def _volume_exists(self):
         ).stdout
         names = text.split('\n')
         return self.name in names
+
+    def before_restore(self):
+        if not self._volume_exists():
```
Comment (richfitz) [inline targets.py:33]: What should we do if the volume already exists?  Overwrite, delete then overwrite, error?

File: docker_rsync.py (hunk 4 above)
Comment (hillalex) [reply docker_rsync.py]: oh neat :+1:

File: docker_rsync.py (hunk 1 above)
Comment (hillalex) [reply docker_rsync.py]: when I run with the user id (obtained with `getuid()`)  I get an error - "No user exists for uid 1003"

File: targets.py (hunk 6 above)
Comment (hillalex) [reply targets.py:33]: At the moment it'll just overwrite, which is the same thing it'll do with a directory. I've set the rsync `--delete` flag so it'll overwrite any changed files and delete any files that don't exist in the backup

File: docker_rsync.py (hunk 1 above)
Comment (richfitz) [reply docker_rsync.py]: Is that coming from python on the host, the docker daemon or from the process in the container itself?

**Review by MartinEden — APPROVED** [review]

**Pull request merged**